File: site/js/synth.js
/* ============================================================================
   Toki — web synthesiser
   ----------------------------------------------------------------------------
   A faithful port of Sources/toki/Synth.swift and Profiles.swift from the macOS
   app. Same model, same parameters, same pre-render architecture:

     transient  band-passed noise, excited 1.5 ms then left to ring — the click
     body       three INHARMONIC modes (1 : 2.71 : 5.18), also band-passed noise
     snap       0.6 ms burst up high — plastic on plastic

   This matters for honesty, not just tidiness: the page claims you are hearing
   the product, so the browser must run the product's engine rather than a
   marketing approximation. If the Swift parameters change, change them here too.

   There are no audio files anywhere in this project. Nothing is fetched.
   ========================================================================== */

export const PROFILES = [
  /* Switches differ in KIND, not just in pitch. A shared band recipe shifted up
     and down makes every voice a sibling — measured: mean timbre distance 1.58,
     with the closest pair at 0.78, which is what "they all sound the same" is in
     numbers. So each profile now carries its own impact structure:

       thud   bottom-out weight — the board taking the hit. Linears live here.
       crack  the high transient. Clickies live here.
       lo/hi  where this switch's impact bands sit relative to its centre —
              a wide spread reads dry and papery, a narrow one reads focused.
       damp   top-end rolloff. High damp = "creamy"; low damp = open and ringy.
  */
  /* Ordered dull -> bright. `name` is the real switch each voice is modelled on
     and `family` is the switch TYPE, which is what a buyer actually chooses by.

     The names stay because they ARE the product's vocabulary: "Cherry MX Blue"
     tells a reader exactly what they are about to hear and an invented name
     tells them nothing. They are used nominatively — to identify what was
     modelled — and every surface that shows them also says these are original
     emulations with no affiliation or endorsement. That pairing is the point:
     naming the reference is not the risk, implying you ARE it would be.

     Slugs are frozen: the synthesis seed is stableHash(slug), so renaming a slug
     would change the sound itself. Names and families are display-only. */
  // slug         name                    family     body bodyHi bodyDec bodyGain  noise  nQ   nDec  nGain   hiHz  hiQ  hiDec hiGain gain  thud crack  lo    hi   damp
  ['cocoa',     'Gateron Milky Yellow', 'Linear', 98, 0.28, 0.072,  0.62,      880, 0.62, 0.030, 1.55,  2600, 0.80, .0060, 0.22, 1.06, 2.10, 0.14, 0.34, 1.35, 0.80],
  ['butter',    'NovelKeys Cream',      'Linear', 152, 0.70, 0.044,  0.38,     1750, 1.15, 0.020, 2.00,  4200, 1.15, .0050, 0.60, 0.99, 1.05, 0.55, 0.55, 1.55, 0.70],
  ['graphite',  'Cherry MX Black',      'Linear', 128, 0.42, 0.052,  0.40,     1350, 0.85, 0.022, 2.30,  3800, 0.95, .0048, 0.55, 1.00, 1.35, 0.45, 0.42, 1.75, 0.45],
  ['vermilion', 'Cherry MX Brown',      'Tactile',178, 0.85, 0.030,  0.26,     2450, 1.00, 0.015, 2.70,  5200, 1.00, .0038, 1.05, 1.00, 0.70, 0.95, 0.48, 2.10, 0.30],
  ['amethyst',  'Gazzew Boba U4T',      'Tactile',205, 1.05, 0.024,  0.16,     3400, 1.35, 0.011, 3.10,  7200, 1.35, .0032, 1.85, 0.95, 0.35, 1.55, 0.62, 2.60, 0.05],
  ['kraft',     'Cherry MX Blue',       'Clicky', 168, 0.62, 0.014,  0.10,     2100, 0.55, 0.008, 3.20,  6000, 0.70, .0026, 1.35, 1.00, 0.22, 1.30, 0.28, 3.10, 0.15],
  ['porcelain', 'Kailh Box Jade',       'Clicky', 232, 1.25, 0.020,  0.14,     4300, 1.50, 0.010, 3.70,  8600, 1.50, .0030, 2.40, 0.92, 0.18, 1.90, 0.70, 3.40, 0.00],
].map(([slug, name, family, body, bodyHi, bodyDec, bodyGain,
        noise, noiseQ, noiseDec, noiseGain, hiHz, hiQ, hiDec, hiGain, gain,
        thud, crack, lo, hi, damp]) =>
  ({ slug, name, family, body, bodyHi, bodyDec, bodyGain,
     noise, noiseQ, noiseDec, noiseGain, hiHz, hiQ, hiDec, hiGain, gain,
     thud, crack, lo, hi, damp,
     /* Peak level per switch. Normalising every voice to one target was the
        largest single homogeniser: a soft linear and a clicky are not the same
        loudness in the world, and making them equal removes the most obvious
        cue that two switches are different. */
     level: ({ cocoa: 0.50, butter: 0.58, graphite: 0.64, vermilion: 0.72,
               amethyst: 0.82, kraft: 0.86, porcelain: 0.98 })[slug] ?? 0.72 }));

export const MODE_RATIOS = [1.0, 2.71, 5.18];
const VARIANTS = 16;
/* Fewer variants in the browser than in the app: seven profiles are cached at
   once here, so 16 each would be 224 buffers and a slow first paint. */
const VARIANTS_WEB = 8;

/* Key size classes — a spacebar is a bigger box with more air in it. */
const SIZE = {
  normal: { dec: 1.00, gain: 1.00, centre: 1.00 },
  wide:   { dec: 1.22, gain: 1.08, centre: 0.94 },
  space:  { dec: 1.45, gain: 1.18, centre: 0.86 },
};

/* RBJ constant-peak-gain band-pass, one biquad, direct form I. */
function bandpass(freq, q, sr) {
  const f = Math.min(Math.max(freq, 20), sr * 0.45);
  const w0 = 2 * Math.PI * f / sr;
  const alpha = Math.sin(w0) / (2 * Math.max(q, 0.05));
  const a0 = 1 + alpha;
  const b0 = alpha / a0, b2 = -alpha / a0;
  const a1 = (-2 * Math.cos(w0)) / a0, a2 = (1 - alpha) / a0;
  let x1 = 0, x2 = 0, y1 = 0, y2 = 0;
  return (x) => {
    const y = b0 * x + b2 * x2 - a1 * y1 - a2 * y2;
    x2 = x1; x1 = x; y2 = y1; y1 = y;
    return y;
  };
}

/* Deterministic PRNG (xorshift64), so a given voice renders identically every
   load. Same reason as the app: with an unseeded source the sound would drift
   between page loads and nothing could be asserted about it. */
function rng(seed) {
  let s = BigInt.asUintN(64, BigInt(seed) || 1n);
  const M = (1n << 64n) - 1n;
  return () => {
    s ^= (s << 13n) & M;
    s ^= s >> 7n;
    s ^= (s << 17n) & M;
    return Number(BigInt.asIntN(64, s)) / 9223372036854775807;
  };
}

function stableHash(str) {
  let h = 0xcbf29ce484222325n;
  for (const ch of str) {
    h ^= BigInt(ch.codePointAt(0));
    h = BigInt.asUintN(64, h * 0x100000001b3n);
  }
  return h;
}

/**
 * Render one hit to a mono Float32Array.
 * `tune` is the live tuner: {bright, body, decay} each ~0.5 at neutral.
 */
export function render(p, { sr = 48000, phase = 'press', size = 'normal',
                            variant = 0, tune = null } = {}) {
  const sz = SIZE[size] || SIZE.normal;
  const seed = stableHash(p.slug) + BigInt(variant) * 7919n
             + (phase === 'press' ? 0n : 104729n)
             + BigInt(size === 'normal' ? 0 : size === 'wide' ? 1301 : 2603);
  const rand = rng(seed);

  const jit = (spread) => 1 + rand() * spread;
  const fJit = jit(0.03), gJit = jit(0.08);

  const rel = phase === 'release';
  const relBody = rel ? 0.35 : 1, relNoise = rel ? 0.55 : 1;
  const relHi = rel ? 0.60 : 1, relDec = rel ? 0.62 : 1, relBright = rel ? 1.15 : 1;

  // Tuner: multiplicative around neutral so the shipped voices are the midpoint.
  const tBright = tune ? 0.55 + tune.bright * 1.35 : 1;   // band centre
  const tBody   = tune ? 0.25 + tune.body   * 1.75 : 1;   // body level
  const tDecay  = tune ? 0.45 + tune.decay  * 1.45 : 1;   // ring length

  const bodyDec  = p.bodyDec * sz.dec * relDec * tDecay;
  const noiseDec = p.noiseDec * relDec * tDecay;
  const noiseF   = p.noise * sz.centre * relBright * fJit * tBright;
  const hiF      = p.hiHz * relBright * fJit * tBright;

  const tail = Math.max(bodyDec * 6, noiseDec * 6) + 0.01;
  const n = Math.floor(tail * sr);
  const out = new Float64Array(n);
  const burst = Math.floor(0.0015 * sr);

  // ---- IMPACT: broadband, not one band -------------------------------------
  // A stem hitting plastic excites a wide spectrum at once. Modelling that as a
  // single band-pass is the main reason v1 read as "filtered noise" rather than
  // "something hit something". Three overlapping bands at different centres and
  // decays sum to an impact: the low one carries weight, the high one carries
  // the crack, and their different decay rates give the spectrum motion as it
  // dies — which is what a recording has and a single band does not.
  const BANDS = [
    { f: p.lo,  q: 0.60, g: 0.90 * (1 + p.damp * 0.5), d: 1.40 },  // weight
    { f: 1.00,  q: 0.95, g: 1.00,                      d: 1.00 },  // the click
    { f: p.hi,  q: 1.25, g: 0.55 * p.crack * (1 - p.damp * 0.75), d: 0.50 },  // crack
  ];
  for (const b of BANDS) {
    if (b.g <= 0.001) continue;
    const bp = bandpass(noiseF * b.f, p.noiseQ * b.q, sr);
    const amp = p.noiseGain * relNoise * gJit * b.g;
    const dec = noiseDec * b.d;
    for (let i = 0; i < n; i++) {
      const v = bp(i < burst ? rand() : 0);
      out[i] += v * amp * Math.exp(-i / (dec * sr));
    }
  }

  // ---- BOTTOM-OUT THUD ------------------------------------------------------
  // The dominant low component of a real keypress, and absent from v1: the whole
  // board takes the impact. Low Q so it thumps instead of ringing a pitch, and a
  // short decay so it reads as weight rather than as a bass note.
  {
    const bp = bandpass(p.body * 0.82, 2.2, sr);
    const amp = p.bodyGain * relBody * 2.6 * p.thud * gJit * tBody;
    const dec = bodyDec * 0.7;
    const hit = Math.floor(0.0022 * sr);
    for (let i = 0; i < n; i++) {
      const v = bp(i < hit ? rand() : 0);
      out[i] += v * amp * Math.exp(-i / (dec * sr));
    }
  }

  // body — inharmonic, because a case is a box and not a string
  MODE_RATIOS.forEach((ratio, k) => {
    const f = p.body * ratio * fJit;
    if (f >= sr * 0.45) return;
    const modeGain = k === 0 ? 1 : p.bodyHi * (k === 1 ? 0.7 : 0.4);
    const modeDec = bodyDec * (k === 0 ? 1 : 0.55);
    // Q lowered from 7+3k: high-Q modes ring like a struck tube and were a
    // large part of the synthetic character. These should tint the impact.
    const bp = bandpass(f, 4.5 + k * 1.8, sr);
    const mrand = rng(9176n + BigInt(k) * 31n + BigInt(variant));
    const amp = p.bodyGain * relBody * modeGain * gJit * tBody;
    for (let i = 0; i < n; i++) {
      const v = bp(i < burst ? mrand() : 0);
      out[i] += v * amp * Math.exp(-i / (modeDec * sr));
    }
  });

  // edge snap
  {
    const bp = bandpass(hiF, p.hiQ, sr);
    const amp = p.hiGain * relHi * gJit;
    const hb = Math.floor(0.0006 * sr);
    for (let i = 0; i < n; i++) {
      const v = bp(i < hb ? rand() : 0);
      out[i] += v * amp * Math.exp(-i / (p.hiDec * sr));
    }
  }

  // shaping: soft clip, de-click both ends, normalise to consistent headroom
  const trim = p.gain * sz.gain;
  const fi = Math.max(Math.floor(0.0002 * sr), 1);
  const fo = Math.max(Math.floor(0.001 * sr), 1);
  let peak = 0;
  for (let i = 0; i < n; i++) {
    let v = Math.tanh(out[i] * trim * 1.4) / 1.4;
    if (i < fi) v *= i / fi;
    if (i > n - fo) v *= (n - i) / fo;
    out[i] = v;
    peak = Math.max(peak, Math.abs(v));
  }
  const norm = peak > 1e-9 ? (p.level ?? 0.72) / peak : 1;
  const f32 = new Float32Array(n);
  for (let i = 0; i < n; i++) f32[i] = out[i] * norm;
  return f32;
}

/** Spectral centroid — used by the page to show how bright a voice is. */
export function centroid(x, sr = 48000) {
  let num = 0, den = 0;
  for (let f = 200; f < sr * 0.45; f *= 1.12) {
    let re = 0, im = 0;
    const w = 2 * Math.PI * f / sr;
    for (let i = 0; i < x.length; i += 2) {   // stride 2: plenty for a readout
      re += x[i] * Math.cos(w * i);
      im += x[i] * Math.sin(w * i);
    }
    const mag = Math.hypot(re, im);
    num += mag * f; den += mag;
  }
  return den > 0 ? num / den : 0;
}

/* ---------------------------------------------------------------------------
   Playback. Mirrors the app: every buffer is rendered up front, so a keypress
   costs one buffer schedule. Voices are chosen PER KEY, not randomly — on a real
   board each key is its own physical switch, so pressing `a` twice sounds the
   same while `a` and `s` differ. Per-press liveliness comes from gain instead.
   --------------------------------------------------------------------------- */
export class Player {
  constructor() {
    this.ctx = null;
    this.master = null;
    /* buffers[slug][phase][size] -> [variant]
       ALL profiles are cached, not just the current one, because the typing test
       changes switch mid-sentence. Rebuilding 96 buffers at a sentence boundary
       would stall the page exactly when the visitor is mid-flow. */
    this.cache = {};
    this.tune = null;
    this.volume = 0.63;
    this.ready = new Set();
    this.onProgress = null;
  }

  async ensure() {
    if (!this.ctx) {
      const AC = window.AudioContext || window.webkitAudioContext;
      if (!AC) return false;
      this.ctx = new AC({ sampleRate: 48000 });
      this.master = this.ctx.createGain();
      this.master.gain.value = this.volume;
      this.master.connect(this.ctx.destination);
    }
    if (this.ctx.state === 'suspended') await this.ctx.resume();
    this.precompute();          // usually already done at page load
    if (this.ready.size < PROFILES.length) this.wrapAll();
    return this.ctx.state === 'running';
  }

  /* Compute raw PCM with no AudioContext. render() is pure maths, so this can
     run at page load, long before the browser will allow audio. On the first
     gesture all that remains is wrapping Float32Arrays into AudioBuffers, which
     is fast — otherwise the first keystroke waits ~300 ms for a render and the
     visitor hears nothing for the one keypress that matters most. */
  precompute() {
    if (this.pcm) return this.pcm;
    this.pcm = {};
    for (const p of PROFILES) {
      const bank = { press: {}, release: {} };
      for (const phase of ['press', 'release']) {
        for (const size of ['normal', 'space']) {
          const list = [];
          for (let v = 0; v < VARIANTS_WEB; v++) {
            list.push(render(p, { sr: 48000, phase, size, variant: v, tune: this.tune }));
          }
          bank[phase][size] = list;
        }
      }
      this.pcm[p.slug] = bank;
    }
    return this.pcm;
  }

  /** Wrap precomputed PCM into AudioBuffers. Cheap; needs a live context. */
  wrapAll() {
    const sr = this.ctx.sampleRate;
    for (const p of PROFILES) {
      const src = this.pcm[p.slug];
      const out = { press: {}, release: {} };
      for (const phase of ['press', 'release']) {
        for (const size of ['normal', 'space']) {
          out[phase][size] = src[phase][size].map((data) => {
            const buf = this.ctx.createBuffer(1, data.length, sr);
            buf.copyToChannel(data, 0);
            return buf;
          });
        }
      }
      this.cache[p.slug] = out;
      this.ready.add(p.slug);
    }
    this.onProgress?.(this.ready.size, PROFILES.length);
  }

  async buildProfile(p) {
    const sr = this.ctx.sampleRate;
    const out = { press: {}, release: {} };
    for (const phase of ['press', 'release']) {
      for (const size of ['normal', 'space']) {
        const list = [];
        for (let v = 0; v < VARIANTS_WEB; v++) {
          const data = render(p, { sr, phase, size, variant: v, tune: this.tune });
          const buf = this.ctx.createBuffer(1, data.length, sr);
          buf.copyToChannel(data, 0);
          list.push(buf);
        }
        out[phase][size] = list;
        await new Promise(r => setTimeout(r, 0));   // keep the page responsive
      }
    }
    this.cache[p.slug] = out;
    this.ready.add(p.slug);
    this.onProgress?.(this.ready.size, PROFILES.length);
  }

  async buildRest() {
    if (this._resting) return;
    this._resting = true;
    for (const p of PROFILES) {
      if (!this.ready.has(p.slug)) await this.buildProfile(p);
    }
    this._resting = false;
  }

  /** Re-render everything for new tuner settings. */
  async setTune(t) {
    this.tune = t;
    this.cache = {}; this.ready.clear(); this.pcm = null;
    this.precompute();
    if (this.ctx) this.wrapAll();
  }

  setVolume(v) { this.volume = v; if (this.master) this.master.gain.value = v; }

  isReady(slug) { return this.ready.has(slug); }

  /** Falls back to the first rendered voice if `slug` has not finished yet. */
  play(slug, keyCode = 0, { phase = 'press', size = 'normal', pan = 0 } = {}) {
    if (!this.ctx) return false;
    const bank = this.cache[slug] || this.cache[PROFILES[0].slug];
    if (!bank) return false;
    const list = bank[phase][size] || bank[phase].normal;
    if (!list?.length) return false;

    const vi = Math.abs(Math.imul(keyCode, 2654435761)) % list.length;
    const src = this.ctx.createBufferSource();
    src.buffer = list[vi];
    const g = this.ctx.createGain();
    g.gain.value = 1 + (Math.random() * 0.14 - 0.07);
    src.connect(g);
    if (this.ctx.createStereoPanner) {
      const pn = this.ctx.createStereoPanner();
      pn.pan.value = Math.max(-1, Math.min(1, pan));
      g.connect(pn); pn.connect(this.master);
    } else {
      g.connect(this.master);
    }
    src.start();
    return true;
  }
}

/* Approximate horizontal position per physical key, for the stereo image. */
const ROWS = ['`1234567890-=', 'qwertyuiop[]\\', "asdfghjkl;'", 'zxcvbnm,./'];
export function panFor(char) {
  const c = (char || '').toLowerCase();
  for (const row of ROWS) {
    const i = row.indexOf(c);
    if (i >= 0) return ((i / (row.length - 1)) * 2 - 1) * 0.55;
  }
  return 0;
}
export function sizeFor(key) { return (key === ' ') ? 'space' : 'normal'; }
export function codeFor(key) {
  let h = 0;
  for (const ch of String(key)) h = (h * 31 + ch.codePointAt(0)) | 0;
  return h;
}

File: app/Sources/toki/Synth.swift
import Foundation

/// Offline synthesis of one keypress. Everything is rendered ahead of time at
/// launch and played back from memory, so a keystroke costs a buffer schedule
/// and nothing else — no filtering or allocation on the audio path.

/// RBJ constant-peak-gain band-pass. One biquad, direct form I.
struct BandPass {
  private var b0 = 0.0, b1 = 0.0, b2 = 0.0, a1 = 0.0, a2 = 0.0
  private var x1 = 0.0, x2 = 0.0, y1 = 0.0, y2 = 0.0

  init(freq: Double, q: Double, sampleRate: Double) {
    // Guard against a centre above Nyquist, which would fold the band down to
    // an audibly wrong frequency instead of failing.
    let f = min(max(freq, 20.0), sampleRate * 0.45)
    let w0 = 2.0 * Double.pi * f / sampleRate
    let alpha = sin(w0) / (2.0 * max(q, 0.05))
    let a0 = 1.0 + alpha
    b0 = alpha / a0
    b1 = 0.0
    b2 = -alpha / a0
    a1 = (-2.0 * cos(w0)) / a0
    a2 = (1.0 - alpha) / a0
  }

  mutating func process(_ x: Double) -> Double {
    let y = b0 * x + b1 * x1 + b2 * x2 - a1 * y1 - a2 * y2
    x2 = x1; x1 = x
    y2 = y1; y1 = y
    return y
  }
}

/// Stable string hash (FNV-1a, 64-bit).
///
/// NOT `String.hashValue`: Swift seeds that randomly per process, so using it to
/// derive a synthesis seed made every switch sound subtly different on every
/// launch. It went unnoticed because the self-test's determinism check compared
/// two renders *within one process*, where the seed is constant — the check was
/// narrower than the claim it printed. Caught by md5-ing the self-test output
/// across two runs and getting different digests.
func stableHash(_ s: String) -> UInt64 {
  var h: UInt64 = 0xcbf29ce484222325
  for b in s.utf8 {
    h ^= UInt64(b)
    h = h &* 0x100000001b3
  }
  return h
}

/// Small deterministic PRNG so a given (profile, phase, variant) always renders
/// identically — across runs, not just within one. Determinism is what makes the
/// self-test meaningful: with a per-process seed, "peak = 0.72" is unreproducible
/// and cannot be asserted against.
struct SeededRandom {
  private var state: UInt64
  init(seed: UInt64) { state = seed == 0 ? 0x9E3779B97F4A7C15 : seed }
  mutating func nextUnit() -> Double {   // white noise in [-1, 1)
    state ^= state << 13
    state ^= state >> 7
    state ^= state << 17
    return Double(Int64(bitPattern: state)) / Double(Int64.max)
  }
}

enum Phase { case press, release }

struct Synth {
  let sampleRate: Double

  /// Render one hit to mono float samples.
  ///
  /// `variant` selects a deterministic jitter set. Pre-rendering several
  /// variants is what produces the randomised-pitch character without doing any
  /// resampling at playback time: a real keyboard never makes the exact same
  /// sound twice, and repeating one buffer verbatim is instantly recognisable as
  /// synthetic.
  /// `normaliseAgainst` supplies the peak to normalise by, instead of this mix's
  /// own peak.
  ///
  /// Without it the voicing does nothing measurable. Normalising each render to
  /// its own peak means a quieter mix is simply scaled back up: measured with an
  /// FFT, the headphone and speaker voicings came out at exactly 100% of each
  /// other's peak, with bass at 85% and treble at 101% — i.e. the cuts were being
  /// undone. Normalising the headphone mix against the SPEAKER peak keeps them.
  func render(profile p: Profile, phase: Phase, size: KeySize, variant: Int,
              output: OutputKind = .speakers,
              normaliseAgainst: Double? = nil,
              peakOnly: UnsafeMutablePointer<Double>? = nil) -> [Float] {
    var rng = SeededRandom(seed: stableHash(p.slug)
                                 &+ UInt64(variant) &* 7919
                                 &+ (phase == .press ? 0 : 104729)
                                 &+ UInt64(sizeSalt(size)))

        // Jitter: +/-3% on frequencies, +/-8% on gains. Enough to break the
    // xerox-copy effect, small enough that the switch keeps its identity.
    let j = { (spread: Double) -> Double in 1.0 + rng.nextUnit() * spread }
    let fJit = j(0.03), gJit = j(0.08)

    // The release is a quieter, brighter, shorter version of the press: the
    // stem hitting the top housing has less cavity behind it than the bottom-out.
    let isRelease = (phase == .release)
    let relBody  = isRelease ? 0.35 : 1.0
    let relNoise = isRelease ? 0.55 : 1.0
    let relHi    = isRelease ? 0.60 : 1.0
    let relDec   = isRelease ? 0.62 : 1.0
    let relBright = isRelease ? 1.15 : 1.0

    let bodyDec  = p.bodyDec * size.bodyDecayScale * relDec
    let noiseDec = p.noiseDec * relDec
    let noiseF   = p.noise * size.noiseCentreScale * relBright * fJit
    let hiF      = p.hiHz * relBright * fJit

    // Long enough for the slowest mode to fall below audibility.
    let tail = max(bodyDec * 6.0, noiseDec * 6.0) + 0.01
    let n = Int(tail * sampleRate)
    var out = [Double](repeating: 0.0, count: n)

    // --- IMPACT: broadband, not one band -------------------------------------
    // A stem hitting plastic excites a wide spectrum at once. Modelling that as a
    // single band-pass is why the first version read as "filtered noise" rather
    // than "something hit something". Three overlapping bands with different
    // centres and decays give the spectrum motion as it dies, which a recording
    // has and a single band cannot. The ratios are PER PROFILE, so a linear and a
    // clicky differ in kind rather than in pitch.
    let burst = Int(0.0015 * sampleRate)
    let bands: [(f: Double, q: Double, g: Double, d: Double)] = [
      (p.lo,  0.60, 0.90 * (1 + p.damp * 0.5) * output.weightScale, 1.40),   // weight
      (1.00,  0.95, 1.00,                      1.00),   // the click
      (p.hi,  1.25, 0.55 * p.crack * (1 - p.damp * 0.75) * output.crackScale, 0.50),  // crack
    ]
    for b in bands where b.g > 0.001 {
      var bp = BandPass(freq: noiseF * b.f, q: p.noiseQ * b.q, sampleRate: sampleRate)
      let amp = p.noiseGain * relNoise * gJit * b.g
      let dec = noiseDec * b.d
      for i in 0..<n {
        let drive = i < burst ? rng.nextUnit() : 0.0
        let v = bp.process(drive)
        out[i] += v * amp * exp(-Double(i) / (dec * sampleRate))
      }
    }

    // --- BOTTOM-OUT THUD -------------------------------------------------------
    // The dominant low component of a real keypress, and missing entirely from
    // the first model: the whole board takes the impact. Low Q so it thumps
    // instead of ringing a pitch.
    do {
      var bp = BandPass(freq: p.body * 0.82, q: 2.2, sampleRate: sampleRate)
      let amp = p.bodyGain * relBody * 2.6 * p.thud * output.thudScale * gJit
      let dec = bodyDec * 0.7
      let hit = Int(0.0022 * sampleRate)
      for i in 0..<n {
        let drive = i < hit ? rng.nextUnit() : 0.0
        let v = bp.process(drive)
        out[i] += v * amp * exp(-Double(i) / (dec * sampleRate))
      }
    }

    // --- body: three inharmonic modes, also band-passed noise ---------------
    // These were oscillators at first, which was wrong: an oscillator puts all
    // its power at one frequency while the noise is spread across a band, so
    // matching them by gain made the body inaudible next to the transient.
    // Exciting band-passed noise instead puts both voices on the same footing.
    for (k, ratio) in Profiles.modeRatios.enumerated() {
      let modeF = p.body * ratio * fJit
      guard modeF < sampleRate * 0.45 else { continue }
      // Upper modes ring relative to the lowest — this is the thock/clack knob.
      let modeGain = (k == 0 ? 1.0 : p.bodyHi * (k == 1 ? 0.7 : 0.4))
      let modeDec = bodyDec * (k == 0 ? 1.0 : 0.55)
      // Q lowered from 7+3k: high-Q modes ring like a struck tube, which was a
      // large part of the synthetic character. These should tint, not sing.
      var bp = BandPass(freq: modeF, q: 4.5 + Double(k) * 1.8, sampleRate: sampleRate)
      var mrng = SeededRandom(seed: UInt64(9176 &+ k &* 31 &+ variant))
      let amp = p.bodyGain * relBody * modeGain * output.bodyScale * gJit
      for i in 0..<n {
        let drive = i < burst ? mrng.nextUnit() : 0.0
        let v = bp.process(drive)
        out[i] += v * amp * exp(-Double(i) / (modeDec * sampleRate))
      }
    }

    // --- edge snap: very short, very high ----------------------------------
    var bpHi = BandPass(freq: hiF, q: p.hiQ, sampleRate: sampleRate)
    let hiAmp = p.hiGain * relHi * output.snapScale * gJit
    let hiBurst = Int(0.0006 * sampleRate)
    for i in 0..<n {
      let drive = i < hiBurst ? rng.nextUnit() : 0.0
      let v = bpHi.process(drive)
      out[i] += v * hiAmp * exp(-Double(i) / (p.hiDec * sampleRate))
    }

    // --- shaping -----------------------------------------------------------
    let trim = p.gain * size.gainScale
    // A 0.2 ms fade-in removes the DC step at t=0. Without it every keystroke
    // carries a thump that is not part of the switch.
    let fadeIn = max(Int(0.0002 * sampleRate), 1)
    // A 1 ms fade-out prevents a discontinuity at buffer end.
    let fadeOut = max(Int(0.001 * sampleRate), 1)

    var peak = 0.0
    for i in 0..<n {
      var v = out[i] * trim
      v = tanh(v * 1.4) / 1.4                       // soft clip, keeps transients honest
      if i < fadeIn { v *= Double(i) / Double(fadeIn) }
      if i > n - fadeOut { v *= Double(n - i) / Double(fadeOut) }
      out[i] = v
      peak = max(peak, abs(v))
    }

    // Per-profile peak, NOT a single shared target. Normalising every voice to
    // one level was the single largest homogeniser — a soft linear and a clicky
    // are not equally loud in the world, and making them equal removes the most
    // obvious cue that two switches are different.
    if let peakOnly { peakOnly.pointee = peak; return [] }
    let target = p.level
    let basis = normaliseAgainst ?? peak
    let norm = basis > 1e-9 ? target / basis : 1.0
    return out.map { Float($0 * norm) }
  }

  /// Pre-normalisation peak of the speaker mix — the reference a voiced render
  /// must normalise against so its cuts survive.
  func speakerPeak(profile p: Profile, phase: Phase, size: KeySize, variant: Int) -> Double {
    var pk = 0.0
    _ = render(profile: p, phase: phase, size: size, variant: variant,
               output: .speakers, peakOnly: &pk)
    return pk
  }

  private func sizeSalt(_ s: KeySize) -> Int {
    switch s { case .normal: return 0; case .wide: return 1301; case .space: return 2603 }
  }
}
